spark/src/main/java/org/apache/iceberg/actions/BaseSparkAction.java

```diff
@@ -128,33 +127,48 @@
     return manifestDF.union(otherMetadataFileDF).union(manifestListDF);
   }
 
+  // Attempt to use Spark3 Catalog resolution if available on the path
+  private static DynMethods.StaticMethod loadCatalogImpl = null;
+
+  private static Dataset<Row> loadCatalogMetadataTable(SparkSession spark, String tableName, MetadataTableType type) {
+    if (loadCatalogImpl == null) {
+      try {
+        loadCatalogImpl = DynMethods.builder("loadCatalogMetadataTable")
+            .hiddenImpl("org.apache.iceberg.spark.Spark3Util",
+                SparkSession.class, String.class, MetadataTableType.class)
+            .buildStaticChecked();
```

Comment:
This doesn't need to load the method each time it is called. Usually, I would add `orNoop` and call `build` to construct a static field. Then in this method, you'd just need to check whether you have the method or noop:

```java
  Preconditions.checkArgument(!LOAD_CATALOG.isNoop(), "Cannot find Spark3Util class ...");
  LOAD_CATALOG.invoke(spark, tableName, type);
```

Reply:
That's nifty, I originally had a second variable to check if it was failed to load but thought that was a bit of a waste. This seems pretty clean though